### PR TITLE
Move battery type enumeration out of parameters object

### PR DESCRIPTION
**Pull request by peterbarker**
I think a an extra cast here and there is worth the significant de-uglification here.

### CODE REVIEW

File: libraries/AP_HAL/board/linux.h [hunk 1]
```diff
@@ -78,7 +78,7 @@
     * 240x240 crop rescaled to 64x64 */
     #define HAL_FLOW_PX4_FOCAL_LENGTH_MILLIPX (2.21 / (3.6 * 2.0 * 240 / 64))
     #define HAL_RANGEFINDER_LIGHTWARE_I2C_BUS 0
-    #define HAL_BATT_MONITOR_DEFAULT AP_BattMonitor_Params::BattMonitor_TYPE_BEBOP
+#define HAL_BATT_MONITOR_DEFAULT AP_BattMonitor::Type::BEBOP
```
Comment (amilcarlucas) [inline libraries/AP_HAL/board/linux.h:81]: Indentation

Comment (peterbarker) [reply libraries/AP_HAL/board/linux.h]: Fixed.

File: libraries/AP_BattMonitor/AP_BattMonitor.cpp [hunk 2]
```diff
@@ -104,7 +104,7 @@ AP_BattMonitor::init()
 #ifdef HAL_BATT_MONITOR_DEFAULT
     if (_params[0]._type == 0) {
         // we can't use set_default() as the type is used as a flag for parameter conversion
-        _params[0]._type.set((AP_BattMonitor_Params::BattMonitor_Type)HAL_BATT_MONITOR_DEFAULT);
+        _params[0]._type.set(float(HAL_BATT_MONITOR_DEFAULT));
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor.cpp:107]: `_type` is `AP_Int8` why cast to float?

Comment (peterbarker) [reply libraries/AP_BattMonitor/AP_BattMonitor.cpp:107]: Can't pass the enumeration type into the `set` method.

We pass both the enumeration type *and* flat values in as that macro value.

Comment (peterbarker) [reply libraries/AP_BattMonitor/AP_BattMonitor.cpp:107]: .... but there's still the question as to "float" rather than Int.  Reflex, probably.

Comment (peterbarker) [reply libraries/AP_BattMonitor/AP_BattMonitor.cpp]: Fixed.

File: libraries/AP_BattMonitor/AP_BattMonitor.cpp [hunk 3]
```diff
@@ -104,7 +104,7 @@ AP_BattMonitor::init()
 #ifdef HAL_BATT_MONITOR_DEFAULT
     if (_params[0]._type == 0) {
         // we can't use set_default() as the type is used as a flag for parameter conversion
-        _params[0]._type.set((AP_BattMonitor_Params::BattMonitor_Type)HAL_BATT_MONITOR_DEFAULT);
+        _params[0]._type.set(uint8_t(HAL_BATT_MONITOR_DEFAULT));
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor.cpp:107]:
```suggestion
        _params[0]._type.set(int8_t(HAL_BATT_MONITOR_DEFAULT));
```

Comment (peterbarker) [reply libraries/AP_BattMonitor/AP_BattMonitor.cpp]: Done.

**Comment by peterbarker** [comment]
Merged as discussed on devcall.

Possible future niceness change would be to have separate backends for the two different analogue types, one inheritting from the other.

**Pull request merged**